### DISCUSSION TITLE
How/when nfs is mounted ?

### DISCUSSION THREAD

**Discussion by kserin** [dracutdevs/dracut#2538] [Q&A]
Hello !

I'm currently trying to boot a machine using PXE and NFS mount.

I've generated a dracut initramfs on my Debian12 using the following command:  `dracut --add-drivers overlay --force /tmp/initrd.img 6.1.0-13-amd64`.
And the PXE kernel parameters are: `initrd=initrd.img ip=dhcp root=nfs:20.0.1.1:/nfs rd.shell rd.debug`

But when I try to boot, I get an initqueue timeout because of the finished/nfsroot.sh script ( https://github.com/dracutdevs/dracut/blob/master/modules.d/95nfs/parse-nfsroot.sh#L124). So I tried to comment the line but the mount does not seem to be done (Logs indicate `Failed to switch root: Specified switch root path '/sysroot' does not seem to be an OS tree` ).

So I tried to figure out how the NFS mount is done and it seems that `nfsroot.sh` script do this job : https://github.com/dracutdevs/dracut/blob/master/modules.d/95nfs/nfsroot.sh#L17 and it is installed in `/sbin` by https://github.com/dracutdevs/dracut/blob/master/modules.d/95nfs/module-setup.sh#L112
But I cannot find where `/sbin/nfsroot` is called. I probably miss something ... Can someone explain it to me ? :)

**Comment by LaszloGombos** [comment]
@kserin perhaps you can try with https://github.com/dracutdevs/dracut/pull/2269 applied (you might need to build dracut yourself from upstream or at least try to get the patches for the overlay dracut modules).

**Reply by kserin** [reply]
@LaszloGombos It seems this commit concern overlayfs right ?
For now, I dont try to use overlay. I know I put `--add-drivers overlay` in dracut command and I might be confusiong. But I did not put any overlay kernel option.
I just try to mount a working NFS.

**Comment by kserin** [comment]
I succeed to get NFS mount with this patch : https://github.com/dracutdevs/dracut/compare/master...kserin:dracut:master .
I don't know if it's correct but I still can't figure out how it could work before

**Comment by aafeijoo-suse** [comment]
> And the PXE kernel parameters are: initrd=initrd.img ip=dhcp root=nfs:20.0.1.1:/nfs rd.shell rd.debug

With `rd.debug` set and without your patch, could you attach  `/run/initramfs/rdsosreport.txt`?

**Reply by kserin** [reply]
This is the `rdsosreport` file without my patch.
[rdsosreport.txt](https://github.com/dracutdevs/dracut/files/13248653/rdsosreport.txt)

**Reply by aafeijoo-suse** [reply]
AFAIK NFS does not work with `systemd-networkd`, try adding the `network-legacy` module to your initramfs.

**Reply by kserin** [reply]
@aafeijoo-suse I tried to add `--add-drivers network-legacy` but it still does not work

**Reply by aafeijoo-suse** [reply]
it's a dracut module, not a kernel module. Try: `--add "network network-legacy"`

**Comment by kserin** [comment]
@LaszloGombos @aafeijoo-suse 
Should I create a pull request ? I don't see how it can work without calling `/sbin/nfsroot` script.

**Reply by LaszloGombos** [reply]
@kserin - we have an NFS test as part of our automation - https://github.com/dracutdevs/dracut/tree/master/test/TEST-20-NFS . I think the NFSv4 part of the test passing on Debian and Ubuntu (without any additional changes).

It would be helpful if you could look at what is the difference between how the test is configured and your setup.
